File: handlers/hostInfo/cpu.go
package hostInfo

import (
	"bufio"
	"fmt"
	"github.com/Sirupsen/logrus"
	"math"
	"os"
	"regexp"
	"runtime"
	"strconv"
	"strings"
)

type CPUCollector struct {
	cadvisor CadvisorAPIClient
}

func (c CPUCollector) getCPUInfoData() []string {
	if runtime.GOOS == "linux" {
		file, err := os.Open("/proc/cpuinfo")
		defer file.Close()
		data := []string{}
		if err != nil {
			logrus.Error(err)
		} else {
			scanner := bufio.NewScanner(file)
			scanner.Split(bufio.ScanLines)
			for scanner.Scan() {
				data = append(data, scanner.Text())
			}
		}
		return data
	}
	return []string{}
}

func (c CPUCollector) getLinuxCPUInfo() map[string]interface{} {
	data := map[string]interface{}{}

	procs := []string{}
	if runtime.GOOS == "linux" {
		fileData := c.getCPUInfoData()
		for _, line := range fileData {
			parts := strings.Split(line, ":")
			if strings.TrimSpace(parts[0]) == "model name" {
				procs = append(procs, strings.TrimSpace(parts[1]))
				pattern := "([0-9\\.]+)\\s?GHz"
				freq := regexp.MustCompile(pattern).FindString(parts[1])
				if freq != "" {
					ghz := strings.TrimSpace(freq[:len(freq)-3])
					if ghz != "" {
						mhz, _ := strconv.ParseFloat(ghz, 64)
						data["mhz"] = mhz * 1000
					}
				}
			}
			if _, ok := data["mhz"]; !ok {
				if strings.TrimSpace(parts[0]) == "cpu MHz" {
					mhz, _ := strconv.ParseFloat(strings.TrimSpace(parts[1]), 64)
					data["mhz"] = mhz
				}
			}
		}
		data["modelName"] = procs[0]
		data["count"] = len(procs)
	}

	return data
}

func (c CPUCollector) getCPUPercentage() map[string]interface{} {
	data := map[string]interface{}{}
	cpuCoresPercentages := []string{}

	stats := c.cadvisor.GetStats()

	if len(stats) >= 2 {
		statLatest := stats[len(stats)-1].(map[string]interface{})
		statPrev := stats[len(stats)-2].(map[string]interface{})

		timeDiff := c.cadvisor.TimestampDiff(InterfaceToString(statLatest["timestamp"]), InterfaceToString(statPrev["timestamp"].(string)))
		latestUsage, _ := getFieldsIfExist(statLatest, "cpu", "usage", "per_cpu_usage")
		prevUsage, _ := getFieldsIfExist(statPrev, "cpu", "usage", "per_cpu_usage")
		for i, cu := range InterfaceToArray(latestUsage) {
			coreUsage := InterfaceToString(cu)
			core, _ := strconv.ParseFloat(coreUsage, 64)
			pu := InterfaceToString(InterfaceToArray(prevUsage)[i])
			prev, _ := strconv.ParseFloat(pu, 64)
			cpuUsage := core - prev
			percentage := (cpuUsage / float64(timeDiff)) * 100
			percentage = percentage * 1000 // round to 3
			if percentage > 100000 {
				percentage = 100
			} else {
				percentage = math.Floor(percentage) / 1000
			}
			cpuCoresPercentages = append(cpuCoresPercentages, strconv.FormatFloat(percentage, 'f', -1, 64))
		}
		data["cpuCoresPercentages"] = cpuCoresPercentages
	}
	return data
}

func (c CPUCollector) getLoadAverage() map[string]interface{} {
	// TODO mock not implemented
	return map[string]interface{}{
		"loadAvg": getLoadAverage(),
	}
}

func (c CPUCollector) GetData() map[string]interface{} {
	data := map[string]interface{}{}

	if runtime.GOOS == "linux" {
		for key, value := range c.getLinuxCPUInfo() {
			data[key] = value
		}
		for key, value := range c.getCPUPercentage() {
			data[key] = value
		}
		for key, value := range c.getLoadAverage() {
			data[key] = value
		}
	}
	return data
}

func (c CPUCollector) KeyName() string {
	return "cpuInfo"
}

func (c CPUCollector) GetLabels(prefix string) map[string]string {
	if _, err := os.Stat("/dev/kvm"); err == nil {
		return map[string]string{
			fmt.Sprintf("%s.%s", prefix, "kvm"): "true",
		}
	}
	return map[string]string{}
}

File: tests/tests/cattle/plugins/host_info/cpu.py
import platform
import os
import math
import re

from tests.cattle.utils import CadvisorAPIClient
from tests.cattle import Config


class CpuCollector(object):
    def __init__(self):
        self.cadvisor = CadvisorAPIClient(Config.cadvisor_ip(),
                                          Config.cadvisor_port())

    def _get_cpuinfo_data(self):
        with open('/proc/cpuinfo') as f:
            data = f.readlines()

        return data

    def _get_linux_cpu_info(self):
        data = {}

        procs = []
        file_data = self._get_cpuinfo_data()
        for line in file_data:
            split_line = line.split(':')
            if split_line[0].strip() == "model name":
                procs.append(split_line[1].strip())
                freq = re.search(r'([0-9\.]+)\s?GHz', split_line[1])
                if freq:
                    data['mhz'] = float(freq.group(1)) * 1000

            if 'mhz' not in data:
                if split_line[0].strip() == "cpu MHz":
                    data['mhz'] = float(split_line[1].strip())

        data['modelName'] = procs[0]
        data['count'] = len(procs)

        return data

    def _get_cpu_percentages(self):
        data = {}
        data['cpuCoresPercentages'] = []

        stats = self.cadvisor.get_stats()

        if len(stats) >= 2:
            stat_latest = stats[-1]
            stat_prev = stats[-2]

            time_diff = self.cadvisor.timestamp_diff(stat_latest['timestamp'],
                                                     stat_prev['timestamp'])

            latest_usage = stat_latest['cpu']['usage']['per_cpu_usage']
            prev_usage = stat_prev['cpu']['usage']['per_cpu_usage']

            for idx, core_usage in enumerate(latest_usage):
                cpu_usage = float(core_usage) - float(prev_usage[idx])
                percentage = (cpu_usage/time_diff) * 100
                percentage = round(percentage, 3)

                if percentage > 100:
                    percentage = math.floor(percentage)

                data['cpuCoresPercentages'].append(percentage)

        return data

    def _get_load_average(self):
        return {'loadAvg': list(os.getloadavg())}

    def key_name(self):
        return "cpuInfo"

    def get_data(self):
        data = {}

        if platform.system() == 'Linux':
            data.update(self._get_linux_cpu_info())
            data.update(self._get_load_average())
            data.update(self._get_cpu_percentages())

        return data

    def get_labels(self, pfx="rancher"):
        if os.path.exists('/dev/kvm'):
            return {".".join([pfx, "kvm"]): "true"}
        else:
            return {}
